src/components/network/layers/arrow-layer-vertex.glsl.js

```diff
@@ -68,23 +83,29 @@ float fetchLineDistance(int point) {
  */
 int findFirstLinePointAfterDistance(float distance) {
   int firstPoint = 0;
-  int lastPoint = instanceLinePointCount - 1;
+  int lastPoint = int(instanceLinePointCount) - 1;
   
-  // variable length loops are not supported in GLSL, instanceLinePointCount is an upper bound that
-  // will never be reached as binary search complexity is in O(log(instanceLinePointCount))
-  for (int i = 0; i < instanceLinePointCount; i++) {
-      if (firstPoint + 1 == lastPoint) {
-          return lastPoint; 
-      }   
-      int middlePoint = (firstPoint + lastPoint) / 2;           
-      float middlePointDistance = fetchLineDistance(middlePoint);      
-      if (middlePointDistance <= distance) {
-         firstPoint = middlePoint;
-      } else {
-         lastPoint = middlePoint;                            
-      }  
-  }   
-} 
+  // variable length loops are not supported in WebGL v1, it needs to be a constant and cannot be like in WebGL v2 an
+  // attribute, so we suppose here that we won't have more that 2^log2MaxPointCount points per line...
+  // 
+  // WARNING!!!!
+  // also, we need to avoid break/return in the for loop even if search complete because with a WebGL1 browser
+  // it is not possible to call texture2D inside a non deterministic piece of code
+  // https://shadertoyunofficial.wordpress.com/2017/11/19/avoiding-compiler-crash-or-endless-compilation 
+  const int log2MaxPointCount = 15;
+  for (int i = 0; i < log2MaxPointCount; i++) {
+      if (firstPoint + 1 != lastPoint) {
+          int middlePoint = (firstPoint + lastPoint) / 2;           
+          float middlePointDistance = fetchLineDistance(middlePoint);      
+          if (middlePointDistance <= distance) {
+             firstPoint = middlePoint;
+          } else {
+             lastPoint = middlePoint;                            
+          }
+      }
+  }
+  return lastPoint; 
+}
```

Comment:
Can you test this loop, is it better ?
```
  const int log2MaxPointCount = 15;
  for (int i = 0; i < log2MaxPointCount; i++) {
      int middlePoint = (firstPoint + lastPoint + 1) / 2;           
      float middlePointDistance = fetchLineDistance(middlePoint);      
      int a = int(step(middlePointDistance, distance));
      int b = 1 - a;
      firstPoint = b * firstPoint + a * middlePoint;
      lastPoint = a * lastPoint + b * middlePoint;                            
  }
```
could also be
```
firstPoint = xx ? firstPoint: middlePoint;
lastPoint = !xx ? lastPoint : middlePoint;
`


Trying to make the code fully branchless as per popular wisdom but I don't know if it's useful

Reply:
```
      int a = int(distance < middlePointDistance);
      int b = 1 - a;
      firstPoint = a * firstPoint + b * middlePoint;
      lastPoint = b * lastPoint + a * middlePoint;                            
```
works too

Reply:
I use
     function onAfterRender() {
+        if (Math.random() > 0.99) {
+            console.log(deck.metrics);
+        }
for quick and dirty fps monitoring the browser console.
Using this, I get about 25fps on a big french network with the real lines. Actually depending on the number of iterations, I'm getting
1 iterations: 30fps
10 iterations: 28fps
100 iterations: 25fps
1000 iterations: 8fps

Reply:
It works but it is less optimized as first if avoid many texture fetch

Reply:
for (int i = 0; i < log2MaxPointCount; i++) {
      if (firstPoint + 1 != lastPoint) {
          int middlePoint = (firstPoint + lastPoint + 1) / 2;           
          float middlePointDistance = fetchLineDistance(middlePoint);      
          int a = int(step(middlePointDistance, distance));
          int b = 1 - a;
          firstPoint = b * firstPoint + a * middlePoint;
          lastPoint = a * lastPoint + b * middlePoint;  
      }
  }

Reply:
Still optimized but with one branch less

Reply:
Keep your "if" gives actually the best performance. I suspect this is because it removes expensive texelFetch calls. Let's keep it as is !

Reply:
So I just keep the fist if or all ?